{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap(a, b, min_length=3):\n",
    "    \"\"\" Return length of longest suffix of 'a' matching a prefix of 'b' that is at least 'min_length'\n",
    "        characters long.  If no such overlap exists, return 0. \"\"\"\n",
    "    start = 0  # start all the way at the left\n",
    "    while True:\n",
    "        start = a.find(b[:min_length], start)  # look for b's suffx in a\n",
    "        if start == -1:  # no more occurrences to right\n",
    "            return 0\n",
    "        # found occurrence; check for full suffix/prefix match\n",
    "        if b.startswith(a[start:]):\n",
    "            return len(a)-start\n",
    "        start += 1  # move just past previous match\n",
    "\n",
    "def scs(ss):\n",
    "    \"\"\" Returns shortest common superstring of given strings, which must be the same length \"\"\"\n",
    "    import itertools\n",
    "    shortest_sup = None\n",
    "    for ssperm in itertools.permutations(ss):\n",
    "        sup = ssperm[0]  # superstring starts as first string\n",
    "        for i in range(len(ss)-1):\n",
    "            # overlap adjacent strings A and B in the permutation\n",
    "            olen = overlap(ssperm[i], ssperm[i+1], min_length=1)\n",
    "            # add non-overlapping portion of B to superstring\n",
    "            sup += ssperm[i+1][olen:]\n",
    "        if shortest_sup is None or len(sup) < len(shortest_sup):\n",
    "            shortest_sup = sup  # found shorter superstring\n",
    "    return shortest_sup  # return shortest\n",
    "\n",
    "def scs_list(ss):\n",
    "    \"\"\" Returns the alphebeticaly sorted list of shortest common superstrings of given strings, which must be the same length \"\"\"\n",
    "    import itertools\n",
    "    shortest_sup = []\n",
    "    shortest_length = 0\n",
    "    for ssperm in itertools.permutations(ss):\n",
    "        sup = ssperm[0]  # superstring starts as first string\n",
    "        for i in range(len(ss)-1):\n",
    "            # overlap adjacent strings A and B in the permutation\n",
    "            olen = overlap(ssperm[i], ssperm[i+1], min_length=1)\n",
    "            # add non-overlapping portion of B to superstring\n",
    "            sup += ssperm[i+1][olen:]\n",
    "        if shortest_length == 0:\n",
    "            shortest_sup.append(sup)\n",
    "            shortest_length = len(sup)\n",
    "        elif len(sup) < shortest_length:\n",
    "            shortest_length = len(sup)\n",
    "            shortest_sup = [sup]\n",
    "        elif len(sup) == shortest_length:\n",
    "            shortest_sup.append(sup)\n",
    "        else:\n",
    "            #simply ignore and move on\n",
    "            None\n",
    "    return sorted(shortest_sup)\n",
    "\n",
    "\n",
    "def test01():\n",
    "    \"\"\"Consider the input strings ABC, BCA, CAB. One shortest common superstring is ABCAB but another is BCABC and another is CABCA.\"\"\"\n",
    "    inputStrings = ['ABC', 'BCA', 'CAB']\n",
    "    shortestCommonSuperstring = scs(inputStrings)\n",
    "    assert shortestCommonSuperstring in ['ABCAB', 'BCABC', 'CABCA']\n",
    "\n",
    "def question1():\n",
    "    \"\"\"What is the length of the shortest common superstring of the following strings? CCT, CTT, TGC, TGG, GAT, ATT\"\"\"\n",
    "    inputStrings = ['CCT', 'CTT', 'TGC', 'TGG', 'GAT', 'ATT']\n",
    "    shortestCommonSuperstring = scs(inputStrings) #== 'CCTTGGATTGC'\n",
    "\n",
    "    #print(\"shortest common superstring\", shortestCommonSuperstring)\n",
    "    print(\"What is the length of the shortest common superstring of the following strings? CCT, CTT, TGC, TGG, GAT, ATT\")\n",
    "    print(len(shortestCommonSuperstring))\n",
    "\n",
    "def example1():\n",
    "    strings = ['ABC', 'BCA', 'CAB']\n",
    "    # Returns just one shortest superstring\n",
    "    assert scs(strings) == 'ABCAB'\n",
    "    # Returns list of all superstrings that are tied for shorest\n",
    "    shortestList = scs_list(strings)\n",
    "    assert shortestList == ['ABCAB', 'BCABC', 'CABCA'], 'found ' + str(shortestList)\n",
    "\n",
    "def example2():\n",
    "    ##from pprint import pprint\n",
    "    strings = ['GAT', 'TAG', 'TCG', 'TGC', 'AAT', 'ATA']\n",
    "    # Returns just one shortest superstring\n",
    "    assert scs(strings) == 'TCGATGCAATAG'\n",
    "    # Returns list of all superstrings that are tied for shorest\n",
    "    shortestList = scs_list(strings)\n",
    "    ##pprint(shortestList)\n",
    "\n",
    "    assert shortestList ==  ['AATAGATCGTGC',\n",
    "                             'AATAGATGCTCG',\n",
    "                             'AATAGTCGATGC',\n",
    "                             'AATCGATAGTGC',\n",
    "                             'AATGCTCGATAG',\n",
    "                             'TCGAATAGATGC',\n",
    "                             'TCGATAGAATGC',\n",
    "                             'TCGATGCAATAG',\n",
    "                             'TGCAATAGATCG',\n",
    "                             'TGCAATCGATAG'], 'found ' + str(shortestList)\n",
    "\n",
    "\n",
    "\n",
    "def question2():\n",
    "    \"\"\"How many different shortest common superstrings are there for the input strings given in the previous question?\n",
    "    Hint 1: You can modify the scs function to keep track of this.\"\"\"\n",
    "    #from pprint import pprint\n",
    "    shortestList = scs_list( ['CCT', 'CTT', 'TGC', 'TGG', 'GAT', 'ATT'] )\n",
    "    #pprint(shortestList)\n",
    "    print(\"How many different shortest common superstrings are there for the input strings given in the previous question?\")\n",
    "    print(len(shortestList))\n",
    "\n",
    "\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            #All the reads are the same length (100 bases) \n",
    "            assert len(seq) == 100\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "overlap_cache = {} # \n",
    "\n",
    "#copied from: http://nbviewer.ipython.org/github/Benlangmead/ads1-notebooks/blob/master/4.02_GreedySCS.ipynb\n",
    "def pick_maximal_overlap(reads, k):\n",
    "    \"\"\" Return a pair of reads from the list with a\n",
    "        maximal suffix/prefix overlap >= k.  Returns\n",
    "        overlap length 0 if there are no such overlaps.\"\"\"\n",
    "    import itertools\n",
    "    reada, readb = None, None\n",
    "    best_olen = 0\n",
    "\n",
    "    for a, b in itertools.permutations(reads, 2):\n",
    "        \n",
    "        if (a,b) in overlap_cache.keys() and overlap_cache[(a,b)] >= k:\n",
    "            overlap_len = overlap_cache[(a,b)]\n",
    "            #print(\"overlap_cache hit for (a,b):\", (a,b), \"min was\", k, \"result was:\", overlap_len, \"cache size:\", len(overlap_cache))\n",
    "            return a, b, overlap_len\n",
    "        else:\n",
    "            olen = overlap(a, b, min_length=k)\n",
    "            #olen = overlap(a, b)\n",
    "            if olen > best_olen:\n",
    "                reada, readb = a, b\n",
    "                best_olen = olen\n",
    "                overlap_cache[(a,b)]=best_olen\n",
    "\n",
    "    return reada, readb, best_olen\n",
    "\n",
    "#copied from: http://nbviewer.ipython.org/github/Benlangmead/ads1-notebooks/blob/master/4.02_GreedySCS.ipynb\n",
    "def greedy_scs(reads, k):\n",
    "    \"\"\" Greedy shortest-common-superstring merge.\n",
    "        Repeat until no edges (overlaps of length >= k)\n",
    "        remain. \"\"\"\n",
    "    lenCacheBefore = len(overlap_cache)\n",
    "    read_a, read_b, olen = pick_maximal_overlap(reads, k)\n",
    "\n",
    "    while olen > 0:\n",
    "        reads.remove(read_a)\n",
    "        reads.remove(read_b)\n",
    "        reads.append(read_a + read_b[olen:])\n",
    "        \n",
    "        read_a, read_b, olen = pick_maximal_overlap(reads, k)\n",
    "    \n",
    "    lenCacheAfter = len(overlap_cache)\n",
    "    #print(\"Cached {0} items during this pass\".format(lenCacheAfter-lenCacheBefore))\n",
    "\n",
    "    return ''.join(reads)\n",
    "\n",
    "def validated_greedy_scs():\n",
    "    res1 = greedy_scs(['ABC', 'BCA', 'CAB'], 2)\n",
    "    #print('res1: ', res1)\n",
    "    assert res1 == 'CABCA'\n",
    "\n",
    "    res2 = greedy_scs(['ABCD', 'CDBC', 'BCDA'], 1)\n",
    "    #print('res2: ', res2)\n",
    "    assert res2 == 'CDBCABCDA'\n",
    "\n",
    "def question3and4():\n",
    "    from datetime import datetime\n",
    "    \n",
    "    reads, qualities = readFastq('ads1_week4_reads.fq')\n",
    "    #print(len(reads))\n",
    "    \n",
    "    for i in range (30, 100):\n",
    "        print(\"timestamp: \", datetime.now())\n",
    "        result = greedy_scs(list(reads), i) #we make a copy of the reads as the greedy_scs modifies the list\n",
    "        print(\"Found result which is {0} bases long for k={1}\".format( len(result), i ) )\n",
    "        \n",
    "        # Hint: the virus genome you are assembling is exactly 15,894 bases long\n",
    "        #assert len(result) == 15894\n",
    "        if len(result) == 15894:\n",
    "            print(\"Question3: \", result.count('A'))\n",
    "            print(\"Question4: \", result.count('T'))\n",
    "            return\n",
    "\n",
    "    for i in range (30, 1, -1):\n",
    "        print(\"timestamp: \", datetime.now())\n",
    "        result = greedy_scs(list(reads), i) #we make a copy of the reads as the greedy_scs modifies the list\n",
    "        print(\"Found result which is {0} bases long for k={1}\".format( len(result), i ) )\n",
    "        \n",
    "        # Hint: the virus genome you are assembling is exactly 15,894 bases long\n",
    "        #assert len(result) == 15894\n",
    "        if len(result) == 15894:\n",
    "            print(\"Question3: \", result.count('A'))\n",
    "            print(\"Question4: \", result.count('T'))\n",
    "            return\n",
    "\n",
    "def main():\n",
    "    test01()\n",
    "    question1()\n",
    "    example1()\n",
    "    example2()\n",
    "    question2()\n",
    "    validated_greedy_scs()\n",
    "\n",
    "    question3and4()\n",
    "    \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
